url4c41254592_topskychen_AuthTopk_tgz-pJ8-quickhull3d_QuickHull3DTestJ8/src/utility/CompareFunction.java

```diff
@@ -0,0 +1,181 @@
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Scanner;
+import java.util.StringTokenizer;
+import Math.MathPoint;
+import Math.MathUtility;
+
+/**
+ * @author chenqian
+ *
+ */
+public class CompareFunction {
+
+	public static double functionL(double w, MathPoint p, MathPoint q){
+		return w * MathUtility.getDistance(p.getX(), p.getY(), q.getX(), q.getY()) + (1 - w) * p.getZ();
+	}
+	
+	public static double functionNL(double w, MathPoint p, MathPoint q){
+		return w * w * MathUtility.getDistance2(p.getX(), p.getY(), q.getX(), q.getY()) + (1 - w) * (1 - w) * p.getZ() * p.getZ(); 
+	}
+	
+	public static int topkQuery(ArrayList<data> datas, int k, double x, double y, double w){
+		MathPoint q = new MathPoint(x, y, 0);
+		ArrayList<tuple> tL = new ArrayList<tuple>();
+		ArrayList<tuple> tNL = new ArrayList<tuple>();
+		HashMap<Integer, Integer> hashmap = new HashMap<Integer, Integer>();
+		for(int i = 0; i < datas.size(); i ++){
+			data d = datas.get(i);
+			double dist = functionL(w, d.p, q);
+			int id = d.id;
+			tL.add(new tuple(dist, id, MathUtility.getDistance(d.p.getX(), d.p.getY(), q.getX(), q.getY()), d.p.getZ()));
+			dist = functionNL(w, d.p, q);
+			id = d.id;
+			tNL.add(new tuple(dist, id, MathUtility.getDistance2(d.p.getX(), d.p.getY(), q.getX(), q.getY()), d.p.getZ() * d.p.getZ()));
+		}
+		int res = 0;
+		Collections.sort(tL);
+		Collections.sort(tNL);
+//		System.out.println("============ tL ============");
+//		for(int i = 0; i < k; i ++){
+//			System.out.println(i + ": " + tL.get(i).dist + ", " + tL.get(i).id + ", " + tL.get(i).w1 + ", " + tL.get(i).w2);
+//		}
+//		System.out.println("============ tNL ============");
+//		for(int i = 0; i < k; i ++){
+//			System.out.println(i + ": " + tNL.get(i).dist + ", " + tNL.get(i).id + ", " + tNL.get(i).w1 + ", " + tNL.get(i).w2);
+//		}
+		for(int i = 0; i < tNL.size(); i ++){
+			hashmap.put(tNL.get(i).id,  i + 1);
+		}
+		for(int i = 0, j = 0; i < tL.size(); i ++){
+			int id = tL.get(i).id;
+			int pos = hashmap.get(id);
+			if(pos > 100)continue;
+			j ++;
+			if(res < pos){
+				res = pos;
+			}
+			System.out.print(pos + " ");
+			if(j == k)break;
+		}System.out.println();
+		return res;
+//		System.out.println(tL.firstKey());
+//		System.out.println(tNL.firstKey());
+	}
+	
+	static int getPos(double ratio, double[] ratios){
+		System.out.println(ratio);
+		for(int i = 0; i < ratios.length; i ++){
+			if(ratio >= ratios[i] && ratio <= ratios[i + 1]){
+				return i;
+			}
+		}
+		return ratios.length;
+	}
+	
+	/**
+	 * @param args
+	 * @throws IOException 
+	 */
+	public static void main(String[] args) throws IOException {
+		// TODO Auto-generated method stub
+		Scanner in = new Scanner(System.in);
+		ArrayList<data> datas = new ArrayList<data>();
+		String file = null;
+		if(args.length > 0){
+			file = args[0];
+		}else{
+			file = in.nextLine(); 
+		}
+		LineNumberReader lr = new LineNumberReader(new FileReader(file));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *lr*. The resource is referenced at lines 99, 100, 107. The resource is closed at line 111. There are other execution paths that do not contain closure statements, for example, when LineNumberReader.readLine throws an exception. Ensure *lr* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).